perceval/converters/cqasm_converter.py

```diff
@@ -0,0 +1,331 @@
+                self._converted_processor.add(target * 2, circuit.copy())
+        else:
+            if gate_name not in _CQASM_2_QUBIT_GATES:
+                raise ConversionUnsupportedFeatureError(
+                    f"Unsupported 2-qubit gate { gate_name }")
+            for target in targets:
+                self._create_2_qubit_gates_from_catalog(
+                    gate_name,
+                    self._num_cnots,
+                    controls[0] * 2,
+                    target * 2,
+                    min(target, controls[0]) * 2,
+                    self._use_postselection)
+
+    def convert(self, ast, use_postselection: bool = True) -> Processor:
+        r"""Convert a cQASM quantum program into a `Processor`.
+
+        :param ast: the AST of a cQASM program
+        :type ast: a Program object, as returned by the cQASM parser
+        :param use_postselection: when True, uses a `postprocessed CNOT`
+        as the last gate. Otherwise, uses only `heralded CNOT`
+        :return: the converted processor
+        """
+        if not type(ast) is cqasm.semantic.Program:
+            raise ConversionSyntaxError(f"cQASM parser error: { ast[0] }")
+
+        self._collect_qubit_list(ast)
+        self._num_cnots = sum(
+            (_cs(s.name) == "CNOT") + 2 * (_cs(s.name) in ["CR", "CRk"])
+            for s in ast.block.statements)
+        self._use_postselection = use_postselection
+
+        qubit_names = [
+            f'{ q }[{ i }]' if i >= 0 else q for (q, i) in self._qubit_list]
+        self._configure_processor(ast, qubit_names=qubit_names)
+        for statement in ast.block.statements:
+            self._convert_statement(statement)
+        self.apply_input_state()
+        return self._converted_processor
+        
+    def convert_string(
+            self,
+            source_string: str,
+            use_postselection: bool = True) -> Processor:
+        r"""Convert a cQASM quantum program into a `Processor`.
+
+        :param source_string: The cQASM program stored in a string
+        :param use_postselection: when True, uses a `postprocessed CNOT`
+        as the last gate. Otherwise, uses only `heralded CNOT`
+        :return: the converted processor
+        """
+        ast = cqasm.Analyzer().analyze_string(source_string)
+        return self.convert(ast, use_postselection)
+
+    def convert_file(
+            self,
+            source_file_name: str,
+            use_postselection: bool = False) -> Processor:
+        r"""Convert a cQASM quantum program into a `Processor`.
+
+        :param source_file_name: The path to the cQASM program to load
+        :param use_postselection: when True, uses a `postprocessed CNOT`
+        as the last gate. Otherwise, uses only `heralded CNOT`
+        :return: the converted processor
+        """
+        ast = cqasm.Analyzer().analyze_file(source_file_name)
+        return self.convert(ast, use_postselection)
+
+    def convert_string_v1(
```

Comment:
The first line of QASM instructions contain information on its `version`. Please refactor this segment of code to allow the user to give as input either `v1` or `v3` type with the same method `convert()` (This should be the only entry point for the user) and not call different methods for different versions of `QASM`

Reply:
My bad, the v3 parser throws a parse error on a v1 file, but the parse error string mentions the version found in the file, so one could recover from this parse error and pass the code to the v1 parser.

Reply:
Yes, I would prefer if the user call has the same signature irrespective of `v1` or `v3`.

Reply:
Ok that's what I did.